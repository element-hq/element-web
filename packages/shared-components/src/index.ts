/*
 * Copyright 2025 New Vector Ltd.
 *
 * SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only OR LicenseRef-Element-Commercial
 * Please see LICENSE files in the repository root for full details.
 */

// Components
export * from "./audio/AudioPlayerView";
export * from "./audio/Clock";
export * from "./audio/PlayPauseButton";
export * from "./audio/SeekBar";
export * from "./avatar/AvatarWithDetails";
export * from "./event-tiles/TextualEventView";
export * from "./message-body/MediaBody";
export * from "./pill-input/Pill";
export * from "./pill-input/PillInput";
export * from "./rich-list/RichItem";
export * from "./rich-list/RichList";
export * from "./utils/Box";
export * from "./utils/Flex";

// Utils
export * from "./utils/i18n"
export * from "./utils/humanize"
export * from "./utils/DateUtils"
export * from "./utils/numbers"

// MVVM
export * from "./ViewWrapper"
export * from "./ViewModel"
export * from "./useViewModel"
export * from "./MockViewModel"

