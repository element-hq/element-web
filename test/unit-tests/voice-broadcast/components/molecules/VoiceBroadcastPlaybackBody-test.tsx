/*
Copyright 2024 New Vector Ltd.
Copyright 2022, 2023 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import React from "react";
import { MatrixClient, MatrixEvent } from "matrix-js-sdk/src/matrix";
import { act, render, RenderResult, screen } from "jest-matrix-react";
import userEvent from "@testing-library/user-event";
import { mocked } from "jest-mock";

import {
    VoiceBroadcastInfoState,
    VoiceBroadcastLiveness,
    VoiceBroadcastPlayback,
    VoiceBroadcastPlaybackBody,
    VoiceBroadcastPlaybackEvent,
    VoiceBroadcastPlaybackState,
} from "../../../../../src/voice-broadcast";
import { filterConsole, stubClient } from "../../../../test-utils";
import { mkVoiceBroadcastInfoStateEvent } from "../../utils/test-utils";
import dis from "../../../../../src/dispatcher/dispatcher";
import { Action } from "../../../../../src/dispatcher/actions";
import { SdkContextClass } from "../../../../../src/contexts/SDKContext";

jest.mock("../../../../../src/dispatcher/dispatcher");

// mock RoomAvatar, because it is doing too much fancy stuff
jest.mock("../../../../../src/components/views/avatars/RoomAvatar", () => ({
    __esModule: true,
    default: jest.fn().mockImplementation(({ room }) => {
        return <div data-testid="room-avatar">room avatar: {room.name}</div>;
    }),
}));

describe("VoiceBroadcastPlaybackBody", () => {
    const userId = "@user:example.com";
    const roomId = "!room:example.com";
    const duration = 23 * 60 + 42; // 23:42
    let client: MatrixClient;
    let infoEvent: MatrixEvent;
    let playback: VoiceBroadcastPlayback;
    let renderResult: RenderResult;

    filterConsole(
        // expected for some tests
        "voice broadcast chunk event to skip to not found",
    );

    beforeAll(() => {
        client = stubClient();
        mocked(client.relations).mockClear();
        mocked(client.relations).mockResolvedValue({ events: [] });

        infoEvent = mkVoiceBroadcastInfoStateEvent(
            roomId,
            VoiceBroadcastInfoState.Stopped,
            userId,
            client.getDeviceId(),
        );
    });

    beforeEach(() => {
        playback = new VoiceBroadcastPlayback(
            infoEvent,
            client,
            SdkContextClass.instance.voiceBroadcastRecordingsStore,
        );
        jest.spyOn(playback, "toggle").mockImplementation(() => Promise.resolve());
        jest.spyOn(playback, "getLiveness");
        jest.spyOn(playback, "getState");
        jest.spyOn(playback, "skipTo");
        jest.spyOn(playback, "durationSeconds", "get").mockReturnValue(duration);
    });

    describe("when rendering a buffering voice broadcast", () => {
        beforeEach(() => {
            mocked(playback.getState).mockReturnValue(VoiceBroadcastPlaybackState.Buffering);
            mocked(playback.getLiveness).mockReturnValue("live");
            renderResult = render(<VoiceBroadcastPlaybackBody playback={playback} />);
        });

        it("should render as expected", () => {
            expect(renderResult.container).toMatchSnapshot();
        });
    });

    describe("when rendering a playing broadcast", () => {
        beforeEach(() => {
            mocked(playback.getState).mockReturnValue(VoiceBroadcastPlaybackState.Playing);
            mocked(playback.getLiveness).mockReturnValue("not-live");
            renderResult = render(<VoiceBroadcastPlaybackBody playback={playback} />);
        });

        it("should render as expected", () => {
            expect(renderResult.container).toMatchSnapshot();
        });

        describe("and being in the middle of the playback", () => {
            beforeEach(() => {
                act(() => {
                    playback.emit(VoiceBroadcastPlaybackEvent.TimesChanged, {
                        duration,
                        position: 10 * 60,
                        timeLeft: duration - 10 * 60,
                    });
                });
            });

            describe("and clicking 30s backward", () => {
                beforeEach(async () => {
                    await act(async () => {
                        await userEvent.click(screen.getByLabelText("30s backward"));
                    });
                });

                it("should seek 30s backward", () => {
                    expect(playback.skipTo).toHaveBeenCalledWith(9 * 60 + 30);
                });
            });

            describe("and clicking 30s forward", () => {
                beforeEach(async () => {
                    await act(async () => {
                        await userEvent.click(screen.getByLabelText("30s forward"));
                    });
                });

                it("should seek 30s forward", () => {
                    expect(playback.skipTo).toHaveBeenCalledWith(10 * 60 + 30);
                });
            });
        });

        describe("and clicking the room name", () => {
            beforeEach(async () => {
                await userEvent.click(screen.getByText("My room"));
            });

            it("should not view the room", () => {
                expect(dis.dispatch).not.toHaveBeenCalled();
            });
        });
    });

    describe("when rendering a playing broadcast in pip mode", () => {
        beforeEach(() => {
            mocked(playback.getState).mockReturnValue(VoiceBroadcastPlaybackState.Playing);
            mocked(playback.getLiveness).mockReturnValue("not-live");
            renderResult = render(<VoiceBroadcastPlaybackBody pip={true} playback={playback} />);
        });

        it("should render as expected", () => {
            expect(renderResult.container).toMatchSnapshot();
        });

        describe("and clicking the room name", () => {
            beforeEach(async () => {
                await userEvent.click(screen.getByText("My room"));
            });

            it("should view the room", () => {
                expect(dis.dispatch).toHaveBeenCalledWith({
                    action: Action.ViewRoom,
                    room_id: roomId,
                    metricsTrigger: undefined,
                });
            });
        });
    });

    describe(`when rendering a stopped broadcast`, () => {
        beforeEach(() => {
            mocked(playback.getState).mockReturnValue(VoiceBroadcastPlaybackState.Stopped);
            mocked(playback.getLiveness).mockReturnValue("not-live");
            renderResult = render(<VoiceBroadcastPlaybackBody playback={playback} />);
        });

        it("should render as expected", () => {
            expect(renderResult.container).toMatchSnapshot();
        });

        describe("and clicking the play button", () => {
            beforeEach(async () => {
                await userEvent.click(renderResult.getByLabelText("play voice broadcast"));
            });

            it("should toggle the recording", () => {
                expect(playback.toggle).toHaveBeenCalled();
            });
        });

        describe("and the times update", () => {
            beforeEach(() => {
                act(() => {
                    playback.emit(VoiceBroadcastPlaybackEvent.TimesChanged, {
                        duration,
                        position: 5 * 60 + 13,
                        timeLeft: 7 * 60 + 5,
                    });
                });
            });

            it("should render the times", async () => {
                expect(await screen.findByText("05:13")).toBeInTheDocument();
                expect(await screen.findByText("-07:05")).toBeInTheDocument();
            });
        });
    });

    describe("when rendering an error broadcast", () => {
        beforeEach(() => {
            mocked(playback.getState).mockReturnValue(VoiceBroadcastPlaybackState.Error);
            renderResult = render(<VoiceBroadcastPlaybackBody playback={playback} />);
        });

        it("should render as expected", () => {
            expect(renderResult.container).toMatchSnapshot();
        });
    });

    describe.each([
        [VoiceBroadcastPlaybackState.Paused, "not-live"],
        [VoiceBroadcastPlaybackState.Playing, "live"],
    ] satisfies [VoiceBroadcastPlaybackState, VoiceBroadcastLiveness][])(
        "when rendering a %s/%s broadcast",
        (state: VoiceBroadcastPlaybackState, liveness: VoiceBroadcastLiveness) => {
            beforeEach(() => {
                mocked(playback.getState).mockReturnValue(state);
                mocked(playback.getLiveness).mockReturnValue(liveness);
                renderResult = render(<VoiceBroadcastPlaybackBody playback={playback} />);
            });

            it("should render as expected", () => {
                expect(renderResult.container).toMatchSnapshot();
            });
        },
    );

    it("when there is a broadcast without sender, it should raise an error", () => {
        infoEvent.sender = null;
        expect(() => {
            render(<VoiceBroadcastPlaybackBody playback={playback} />);
        }).toThrow(`Voice Broadcast sender not found (event ${playback.infoEvent.getId()})`);
    });
});
