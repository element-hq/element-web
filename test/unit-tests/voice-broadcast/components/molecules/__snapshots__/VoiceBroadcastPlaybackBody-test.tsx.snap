// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VoiceBroadcastPlaybackBody when rendering a buffering voice broadcast should render as expected 1`] = `
<div>
  <div
    class="mx_VoiceBroadcastBody"
  >
    <div
      class="mx_VoiceBroadcastHeader"
    >
      <div
        data-testid="room-avatar"
      >
        room avatar: 
        My room
      </div>
      <div
        class="mx_VoiceBroadcastHeader_content"
      >
        <div
          class="mx_VoiceBroadcastHeader_room_wrapper"
        >
          <div
            class="mx_VoiceBroadcastHeader_room"
          >
            My room
          </div>
        </div>
        <div
          aria-label="Change input device"
          class="mx_AccessibleButton mx_VoiceBroadcastHeader_line"
          role="button"
          tabindex="0"
        >
          <svg
            class="mx_Icon mx_Icon_16"
            fill="currentColor"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 6a4 4 0 1 1 8 0v6a4 4 0 0 1-8 0V6Z"
            />
            <path
              d="M5 11a1 1 0 0 1 1 1 6 6 0 0 0 12 0 1 1 0 1 1 2 0 8.001 8.001 0 0 1-7 7.938V21a1 1 0 1 1-2 0v-1.062A8.001 8.001 0 0 1 4 12a1 1 0 0 1 1-1Z"
            />
          </svg>
          <span>
            @user:example.com
          </span>
        </div>
        <div
          class="mx_VoiceBroadcastHeader_line"
        >
          <div
            class="mx_Spinner"
          >
            <div
              aria-label="Loading…"
              class="mx_Spinner_icon"
              data-testid="spinner"
              role="progressbar"
              style="width: 14px; height: 14px;"
            />
          </div>
          Buffering…
        </div>
      </div>
      <div
        class="mx_LiveBadge"
      >
        <div
          class="mx_Icon mx_Icon_16"
        />
        Live
      </div>
    </div>
    <div
      class="mx_VoiceBroadcastBody_controls"
    >
      <div
        aria-label="30s backward"
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_secondary_content"
        role="button"
        tabindex="0"
      >
        <div
          class="mx_Icon mx_Icon_24"
        />
      </div>
      <div
        aria-label="pause voice broadcast"
        class="mx_AccessibleButton mx_VoiceBroadcastControl"
        role="button"
        tabindex="0"
      >
        <svg
          class="mx_Icon mx_Icon_12"
          fill="currentColor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 4a2 2 0 0 0-2 2v12a2 2 0 1 0 4 0V6a2 2 0 0 0-2-2Zm8 0a2 2 0 0 0-2 2v12a2 2 0 1 0 4 0V6a2 2 0 0 0-2-2Z"
          />
        </svg>
      </div>
      <div
        aria-label="30s forward"
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_secondary_content"
        role="button"
        tabindex="0"
      >
        <div
          class="mx_Icon mx_Icon_24"
        />
      </div>
    </div>
    <input
      aria-label="Audio seek bar"
      class="mx_SeekBar"
      max="1"
      min="0"
      step="0.001"
      style="--fillTo: 0;"
      tabindex="0"
      type="range"
      value="0"
    />
    <div
      class="mx_VoiceBroadcastBody_timerow"
    >
      <time
        class="mx_Clock"
        datetime="PT0S"
      >
        00:00
      </time>
      <time
        class="mx_Clock"
        datetime="-PT23M42S"
      >
        -23:42
      </time>
    </div>
  </div>
</div>
`;

exports[`VoiceBroadcastPlaybackBody when rendering a pause/not-live broadcast should render as expected 1`] = `
<div>
  <div
    class="mx_VoiceBroadcastBody"
  >
    <div
      class="mx_VoiceBroadcastHeader"
    >
      <div
        data-testid="room-avatar"
      >
        room avatar: 
        My room
      </div>
      <div
        class="mx_VoiceBroadcastHeader_content"
      >
        <div
          class="mx_VoiceBroadcastHeader_room_wrapper"
        >
          <div
            class="mx_VoiceBroadcastHeader_room"
          >
            My room
          </div>
        </div>
        <div
          aria-label="Change input device"
          class="mx_AccessibleButton mx_VoiceBroadcastHeader_line"
          role="button"
          tabindex="0"
        >
          <svg
            class="mx_Icon mx_Icon_16"
            fill="currentColor"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 6a4 4 0 1 1 8 0v6a4 4 0 0 1-8 0V6Z"
            />
            <path
              d="M5 11a1 1 0 0 1 1 1 6 6 0 0 0 12 0 1 1 0 1 1 2 0 8.001 8.001 0 0 1-7 7.938V21a1 1 0 1 1-2 0v-1.062A8.001 8.001 0 0 1 4 12a1 1 0 0 1 1-1Z"
            />
          </svg>
          <span>
            @user:example.com
          </span>
        </div>
        <div
          class="mx_VoiceBroadcastHeader_line"
        >
          <div
            class="mx_Icon mx_Icon_16"
          />
          Voice broadcast
        </div>
      </div>
    </div>
    <div
      class="mx_VoiceBroadcastBody_controls"
    >
      <div
        aria-label="30s backward"
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_secondary_content"
        role="button"
        tabindex="0"
      >
        <div
          class="mx_Icon mx_Icon_24"
        />
      </div>
      <div
        aria-label="resume voice broadcast"
        class="mx_AccessibleButton mx_VoiceBroadcastControl mx_VoiceBroadcastControl-play"
        role="button"
        tabindex="0"
      >
        <svg
          class="mx_Icon mx_Icon_16"
          fill="currentColor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m8.98 4.677 9.921 5.58c1.36.764 1.36 2.722 0 3.486l-9.92 5.58C7.647 20.073 6 19.11 6 17.58V6.42c0-1.53 1.647-2.493 2.98-1.743Z"
          />
        </svg>
      </div>
      <div
        aria-label="30s forward"
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_secondary_content"
        role="button"
        tabindex="0"
      >
        <div
          class="mx_Icon mx_Icon_24"
        />
      </div>
    </div>
    <input
      aria-label="Audio seek bar"
      class="mx_SeekBar"
      max="1"
      min="0"
      step="0.001"
      style="--fillTo: 0;"
      tabindex="0"
      type="range"
      value="0"
    />
    <div
      class="mx_VoiceBroadcastBody_timerow"
    >
      <time
        class="mx_Clock"
        datetime="PT0S"
      >
        00:00
      </time>
      <time
        class="mx_Clock"
        datetime="-PT23M42S"
      >
        -23:42
      </time>
    </div>
  </div>
</div>
`;

exports[`VoiceBroadcastPlaybackBody when rendering a playing broadcast in pip mode should render as expected 1`] = `
<div>
  <div
    class="mx_VoiceBroadcastBody mx_VoiceBroadcastBody--pip"
  >
    <div
      class="mx_VoiceBroadcastHeader"
    >
      <div
        class="mx_AccessibleButton"
        role="button"
        tabindex="0"
      >
        <div
          data-testid="room-avatar"
        >
          room avatar: 
          My room
        </div>
      </div>
      <div
        class="mx_VoiceBroadcastHeader_content"
      >
        <div
          class="mx_AccessibleButton"
          role="button"
          tabindex="0"
        >
          <div
            class="mx_VoiceBroadcastHeader_room_wrapper"
          >
            <div
              class="mx_VoiceBroadcastHeader_room"
            >
              My room
            </div>
          </div>
        </div>
        <div
          aria-label="Change input device"
          class="mx_AccessibleButton mx_VoiceBroadcastHeader_line"
          role="button"
          tabindex="0"
        >
          <svg
            class="mx_Icon mx_Icon_16"
            fill="currentColor"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 6a4 4 0 1 1 8 0v6a4 4 0 0 1-8 0V6Z"
            />
            <path
              d="M5 11a1 1 0 0 1 1 1 6 6 0 0 0 12 0 1 1 0 1 1 2 0 8.001 8.001 0 0 1-7 7.938V21a1 1 0 1 1-2 0v-1.062A8.001 8.001 0 0 1 4 12a1 1 0 0 1 1-1Z"
            />
          </svg>
          <span>
            @user:example.com
          </span>
        </div>
        <div
          class="mx_VoiceBroadcastHeader_line"
        >
          <div
            class="mx_Icon mx_Icon_16"
          />
          Voice broadcast
        </div>
      </div>
    </div>
    <div
      class="mx_VoiceBroadcastBody_controls"
    >
      <div
        aria-label="30s backward"
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_secondary_content"
        role="button"
        tabindex="0"
      >
        <div
          class="mx_Icon mx_Icon_24"
        />
      </div>
      <div
        aria-label="pause voice broadcast"
        class="mx_AccessibleButton mx_VoiceBroadcastControl"
        role="button"
        tabindex="0"
      >
        <svg
          class="mx_Icon mx_Icon_12"
          fill="currentColor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 4a2 2 0 0 0-2 2v12a2 2 0 1 0 4 0V6a2 2 0 0 0-2-2Zm8 0a2 2 0 0 0-2 2v12a2 2 0 1 0 4 0V6a2 2 0 0 0-2-2Z"
          />
        </svg>
      </div>
      <div
        aria-label="30s forward"
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_secondary_content"
        role="button"
        tabindex="0"
      >
        <div
          class="mx_Icon mx_Icon_24"
        />
      </div>
    </div>
    <input
      aria-label="Audio seek bar"
      class="mx_SeekBar"
      max="1"
      min="0"
      step="0.001"
      style="--fillTo: 0;"
      tabindex="0"
      type="range"
      value="0"
    />
    <div
      class="mx_VoiceBroadcastBody_timerow"
    >
      <time
        class="mx_Clock"
        datetime="PT0S"
      >
        00:00
      </time>
      <time
        class="mx_Clock"
        datetime="-PT23M42S"
      >
        -23:42
      </time>
    </div>
  </div>
</div>
`;

exports[`VoiceBroadcastPlaybackBody when rendering a playing broadcast should render as expected 1`] = `
<div>
  <div
    class="mx_VoiceBroadcastBody"
  >
    <div
      class="mx_VoiceBroadcastHeader"
    >
      <div
        data-testid="room-avatar"
      >
        room avatar: 
        My room
      </div>
      <div
        class="mx_VoiceBroadcastHeader_content"
      >
        <div
          class="mx_VoiceBroadcastHeader_room_wrapper"
        >
          <div
            class="mx_VoiceBroadcastHeader_room"
          >
            My room
          </div>
        </div>
        <div
          aria-label="Change input device"
          class="mx_AccessibleButton mx_VoiceBroadcastHeader_line"
          role="button"
          tabindex="0"
        >
          <svg
            class="mx_Icon mx_Icon_16"
            fill="currentColor"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 6a4 4 0 1 1 8 0v6a4 4 0 0 1-8 0V6Z"
            />
            <path
              d="M5 11a1 1 0 0 1 1 1 6 6 0 0 0 12 0 1 1 0 1 1 2 0 8.001 8.001 0 0 1-7 7.938V21a1 1 0 1 1-2 0v-1.062A8.001 8.001 0 0 1 4 12a1 1 0 0 1 1-1Z"
            />
          </svg>
          <span>
            @user:example.com
          </span>
        </div>
        <div
          class="mx_VoiceBroadcastHeader_line"
        >
          <div
            class="mx_Icon mx_Icon_16"
          />
          Voice broadcast
        </div>
      </div>
    </div>
    <div
      class="mx_VoiceBroadcastBody_controls"
    >
      <div
        aria-label="30s backward"
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_secondary_content"
        role="button"
        tabindex="0"
      >
        <div
          class="mx_Icon mx_Icon_24"
        />
      </div>
      <div
        aria-label="pause voice broadcast"
        class="mx_AccessibleButton mx_VoiceBroadcastControl"
        role="button"
        tabindex="0"
      >
        <svg
          class="mx_Icon mx_Icon_12"
          fill="currentColor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 4a2 2 0 0 0-2 2v12a2 2 0 1 0 4 0V6a2 2 0 0 0-2-2Zm8 0a2 2 0 0 0-2 2v12a2 2 0 1 0 4 0V6a2 2 0 0 0-2-2Z"
          />
        </svg>
      </div>
      <div
        aria-label="30s forward"
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_secondary_content"
        role="button"
        tabindex="0"
      >
        <div
          class="mx_Icon mx_Icon_24"
        />
      </div>
    </div>
    <input
      aria-label="Audio seek bar"
      class="mx_SeekBar"
      max="1"
      min="0"
      step="0.001"
      style="--fillTo: 0;"
      tabindex="0"
      type="range"
      value="0"
    />
    <div
      class="mx_VoiceBroadcastBody_timerow"
    >
      <time
        class="mx_Clock"
        datetime="PT0S"
      >
        00:00
      </time>
      <time
        class="mx_Clock"
        datetime="-PT23M42S"
      >
        -23:42
      </time>
    </div>
  </div>
</div>
`;

exports[`VoiceBroadcastPlaybackBody when rendering a playing/live broadcast should render as expected 1`] = `
<div>
  <div
    class="mx_VoiceBroadcastBody"
  >
    <div
      class="mx_VoiceBroadcastHeader"
    >
      <div
        data-testid="room-avatar"
      >
        room avatar: 
        My room
      </div>
      <div
        class="mx_VoiceBroadcastHeader_content"
      >
        <div
          class="mx_VoiceBroadcastHeader_room_wrapper"
        >
          <div
            class="mx_VoiceBroadcastHeader_room"
          >
            My room
          </div>
        </div>
        <div
          aria-label="Change input device"
          class="mx_AccessibleButton mx_VoiceBroadcastHeader_line"
          role="button"
          tabindex="0"
        >
          <svg
            class="mx_Icon mx_Icon_16"
            fill="currentColor"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 6a4 4 0 1 1 8 0v6a4 4 0 0 1-8 0V6Z"
            />
            <path
              d="M5 11a1 1 0 0 1 1 1 6 6 0 0 0 12 0 1 1 0 1 1 2 0 8.001 8.001 0 0 1-7 7.938V21a1 1 0 1 1-2 0v-1.062A8.001 8.001 0 0 1 4 12a1 1 0 0 1 1-1Z"
            />
          </svg>
          <span>
            @user:example.com
          </span>
        </div>
        <div
          class="mx_VoiceBroadcastHeader_line"
        >
          <div
            class="mx_Icon mx_Icon_16"
          />
          Voice broadcast
        </div>
      </div>
      <div
        class="mx_LiveBadge"
      >
        <div
          class="mx_Icon mx_Icon_16"
        />
        Live
      </div>
    </div>
    <div
      class="mx_VoiceBroadcastBody_controls"
    >
      <div
        aria-label="30s backward"
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_secondary_content"
        role="button"
        tabindex="0"
      >
        <div
          class="mx_Icon mx_Icon_24"
        />
      </div>
      <div
        aria-label="pause voice broadcast"
        class="mx_AccessibleButton mx_VoiceBroadcastControl"
        role="button"
        tabindex="0"
      >
        <svg
          class="mx_Icon mx_Icon_12"
          fill="currentColor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M8 4a2 2 0 0 0-2 2v12a2 2 0 1 0 4 0V6a2 2 0 0 0-2-2Zm8 0a2 2 0 0 0-2 2v12a2 2 0 1 0 4 0V6a2 2 0 0 0-2-2Z"
          />
        </svg>
      </div>
      <div
        aria-label="30s forward"
        class="mx_AccessibleButton mx_AccessibleButton_hasKind mx_AccessibleButton_kind_secondary_content"
        role="button"
        tabindex="0"
      >
        <div
          class="mx_Icon mx_Icon_24"
        />
      </div>
    </div>
    <input
      aria-label="Audio seek bar"
      class="mx_SeekBar"
      max="1"
      min="0"
      step="0.001"
      style="--fillTo: 0;"
      tabindex="0"
      type="range"
      value="0"
    />
    <div
      class="mx_VoiceBroadcastBody_timerow"
    >
      <time
        class="mx_Clock"
        datetime="PT0S"
      >
        00:00
      </time>
      <time
        class="mx_Clock"
        datetime="-PT23M42S"
      >
        -23:42
      </time>
    </div>
  </div>
</div>
`;

exports[`VoiceBroadcastPlaybackBody when rendering a stopped broadcast should render as expected 1`] = `
<div>
  <div
    class="mx_VoiceBroadcastBody"
  >
    <div
      class="mx_VoiceBroadcastHeader"
    >
      <div
        data-testid="room-avatar"
      >
        room avatar: 
        My room
      </div>
      <div
        class="mx_VoiceBroadcastHeader_content"
      >
        <div
          class="mx_VoiceBroadcastHeader_room_wrapper"
        >
          <div
            class="mx_VoiceBroadcastHeader_room"
          >
            My room
          </div>
        </div>
        <div
          aria-label="Change input device"
          class="mx_AccessibleButton mx_VoiceBroadcastHeader_line"
          role="button"
          tabindex="0"
        >
          <svg
            class="mx_Icon mx_Icon_16"
            fill="currentColor"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 6a4 4 0 1 1 8 0v6a4 4 0 0 1-8 0V6Z"
            />
            <path
              d="M5 11a1 1 0 0 1 1 1 6 6 0 0 0 12 0 1 1 0 1 1 2 0 8.001 8.001 0 0 1-7 7.938V21a1 1 0 1 1-2 0v-1.062A8.001 8.001 0 0 1 4 12a1 1 0 0 1 1-1Z"
            />
          </svg>
          <span>
            @user:example.com
          </span>
        </div>
        <div
          class="mx_VoiceBroadcastHeader_line"
        >
          <div
            class="mx_Icon mx_Icon_16"
          />
          Voice broadcast
        </div>
      </div>
    </div>
    <div
      class="mx_VoiceBroadcastBody_controls"
    >
      <div
        aria-label="play voice broadcast"
        class="mx_AccessibleButton mx_VoiceBroadcastControl mx_VoiceBroadcastControl-play"
        role="button"
        tabindex="0"
      >
        <svg
          class="mx_Icon mx_Icon_16"
          fill="currentColor"
          height="1em"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m8.98 4.677 9.921 5.58c1.36.764 1.36 2.722 0 3.486l-9.92 5.58C7.647 20.073 6 19.11 6 17.58V6.42c0-1.53 1.647-2.493 2.98-1.743Z"
          />
        </svg>
      </div>
    </div>
    <input
      aria-label="Audio seek bar"
      class="mx_SeekBar"
      max="1"
      min="0"
      step="0.001"
      style="--fillTo: 0;"
      tabindex="0"
      type="range"
      value="0"
    />
    <div
      class="mx_VoiceBroadcastBody_timerow"
    >
      <time
        class="mx_Clock"
        datetime="PT0S"
      >
        00:00
      </time>
      <time
        class="mx_Clock"
        datetime="-PT23M42S"
      >
        -23:42
      </time>
    </div>
  </div>
</div>
`;

exports[`VoiceBroadcastPlaybackBody when rendering an error broadcast should render as expected 1`] = `
<div>
  <div
    class="mx_VoiceBroadcastBody"
  >
    <div
      class="mx_VoiceBroadcastHeader"
    >
      <div
        data-testid="room-avatar"
      >
        room avatar: 
        My room
      </div>
      <div
        class="mx_VoiceBroadcastHeader_content"
      >
        <div
          class="mx_VoiceBroadcastHeader_room_wrapper"
        >
          <div
            class="mx_VoiceBroadcastHeader_room"
          >
            My room
          </div>
        </div>
        <div
          aria-label="Change input device"
          class="mx_AccessibleButton mx_VoiceBroadcastHeader_line"
          role="button"
          tabindex="0"
        >
          <svg
            class="mx_Icon mx_Icon_16"
            fill="currentColor"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M8 6a4 4 0 1 1 8 0v6a4 4 0 0 1-8 0V6Z"
            />
            <path
              d="M5 11a1 1 0 0 1 1 1 6 6 0 0 0 12 0 1 1 0 1 1 2 0 8.001 8.001 0 0 1-7 7.938V21a1 1 0 1 1-2 0v-1.062A8.001 8.001 0 0 1 4 12a1 1 0 0 1 1-1Z"
            />
          </svg>
          <span>
            @user:example.com
          </span>
        </div>
        <div
          class="mx_VoiceBroadcastHeader_line"
        >
          <div
            class="mx_Icon mx_Icon_16"
          />
          Voice broadcast
        </div>
      </div>
    </div>
    <div
      class="mx_VoiceBroadcastRecordingConnectionError"
    >
      <svg
        class="mx_Icon mx_Icon_16"
        fill="currentColor"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M12.713 17.713A.968.968 0 0 1 12 18a.968.968 0 0 1-.713-.287A.967.967 0 0 1 11 17a.97.97 0 0 1 .287-.712A.968.968 0 0 1 12 16a.97.97 0 0 1 .713.288A.968.968 0 0 1 13 17a.97.97 0 0 1-.287.713Zm0-4A.968.968 0 0 1 12 14a.968.968 0 0 1-.713-.287A.967.967 0 0 1 11 13V9a.97.97 0 0 1 .287-.712A.968.968 0 0 1 12 8a.97.97 0 0 1 .713.288A.968.968 0 0 1 13 9v4a.97.97 0 0 1-.287.713Z"
        />
        <path
          clip-rule="evenodd"
          d="M10.264 3.039c.767-1.344 2.705-1.344 3.472 0l8.554 14.969c.762 1.333-.2 2.992-1.736 2.992H3.446c-1.535 0-2.498-1.659-1.736-2.992l8.553-14.969ZM3.446 19 12 4.031l8.554 14.97H3.446Z"
          fill-rule="evenodd"
        />
      </svg>
      Unable to play this voice broadcast
    </div>
  </div>
</div>
`;
